File: nodejs/services/TranslationService.js
/**
 * Translation Service
 * Handles server-side translation loading and parameter substitution
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

class TranslationService {
  static translations = {};

  /**
   * Load translation file for a locale
   * @param {string} locale - Locale code
   * @returns {Object} Translation dictionary
   */
  static loadTranslations(locale) {
    if (this.translations[locale]) {
      return this.translations[locale];
    }

    try {
      const filePath = path.join(__dirname, '..', 'translations', `${locale}.json`);
      const data = fs.readFileSync(filePath, 'utf8');
      this.translations[locale] = JSON.parse(data);
      return this.translations[locale];
    } catch (error) {
      console.error(`Failed to load translations for ${locale}:`, error.message);
      // Fallback to English
      if (locale !== 'en') {
        return this.loadTranslations('en');
      }
      return {};
    }
  }

  /**
   * Get all translations for a locale
   * @param {string} locale - Locale code
   * @returns {Object} Translation dictionary
   */
  static getAllTranslations(locale) {
    return this.loadTranslations(locale);
  }

  /**
   * Translate a key with optional parameters
   * @param {string} key - Translation key
   * @param {string} locale - Locale code
   * @param {Object} params - Parameters to replace (e.g., {amount: '$10.00'})
   * @returns {string} Translated text
   */
  static translate(key, locale, params = {}) {
    const translations = this.loadTranslations(locale);
    let text = translations[key];

    // Fallback to English if translation not found
    if (!text && locale !== 'en') {
      const enTranslations = this.loadTranslations('en');
      text = enTranslations[key] || key;
    } else if (!text) {
      text = key;
    }

    // Replace parameters in the format %paramName%
    Object.keys(params).forEach(param => {
      const regex = new RegExp(`%${param}%`, 'g');
      text = text.replace(regex, params[param]);
    });

    return text;
  }

  /**
   * Get translation with alias 't'
   * @param {string} key - Translation key
   * @param {string} locale - Locale code
   * @param {Object} params - Parameters to replace
   * @returns {string} Translated text
   */
  static t(key, locale, params = {}) {
    return this.translate(key, locale, params);
  }
}

export default TranslationService;

File: dotnet/Services/TranslationService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace GlobalPayments.Services
{
    /// <summary>
    /// Translation Service
    /// Handles server-side translation loading and parameter substitution
    /// </summary>
    public class TranslationService
    {
        private static readonly Dictionary<string, Dictionary<string, string>> Translations = new();

        /// <summary>
        /// Load translation file for a locale
        /// </summary>
        public static Dictionary<string, string> LoadTranslations(string locale)
        {
            if (Translations.ContainsKey(locale))
            {
                return Translations[locale];
            }

            try
            {
                string path = Path.Combine(Directory.GetCurrentDirectory(), "translations", $"{locale}.json");

                if (!File.Exists(path))
                {
                    Console.WriteLine($"Translation file not found: {path}");
                    // Fallback to English
                    if (locale != "en")
                    {
                        return LoadTranslations("en");
                    }
                    return new Dictionary<string, string>();
                }

                string json = File.ReadAllText(path);
                var localeTranslations = JsonSerializer.Deserialize<Dictionary<string, string>>(json);

                Translations[locale] = localeTranslations ?? new Dictionary<string, string>();
                return Translations[locale];
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to load translations for {locale}: {ex.Message}");
                // Fallback to English
                if (locale != "en")
                {
                    return LoadTranslations("en");
                }
                return new Dictionary<string, string>();
            }
        }

        /// <summary>
        /// Get all translations for a locale
        /// </summary>
        public static Dictionary<string, string> GetAllTranslations(string locale)
        {
            return LoadTranslations(locale);
        }

        /// <summary>
        /// Translate a key with optional parameters
        /// </summary>
        public static string Translate(string key, string locale, Dictionary<string, string> parameters = null)
        {
            var localeTranslations = LoadTranslations(locale);
            string text = localeTranslations.GetValueOrDefault(key);

            // Fallback to English if translation not found
            if (text == null && locale != "en")
            {
                var enTranslations = LoadTranslations("en");
                text = enTranslations.GetValueOrDefault(key) ?? key;
            }
            else if (text == null)
            {
                text = key;
            }

            // Replace parameters in the format %paramName%
            if (parameters != null)
            {
                foreach (var param in parameters)
                {
                    text = text.Replace($"%{param.Key}%", param.Value);
                }
            }

            return text;
        }

        /// <summary>
        /// Get translation with alias 't'
        /// </summary>
        public static string T(string key, string locale, Dictionary<string, string> parameters = null)
        {
            return Translate(key, locale, parameters);
        }
    }
}
